Fix the previous correction where we zero'ed out the center coefficient

asalmgren opened the pull request:
of the nodal stencil if it had the wrong sign.  Now we only zero it
if mglev > 0.

## Summary

## Additional background

## Checklist

The proposed changes:
- [X] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

asalmgren committed:
Fix the previous correction where we zero'ed out the center coefficient
of the nodal stencil if it had the wrong sign.  Now we only zero it
if mglev > 0.

asalmgren committed:
Remove trailing white spaces

WeiqunZhang commented:
Does IAMR have positive diagonals on mglev == 0 (i.e., original grids)?

asalmgren commented:
Yes -- and setting sten(i,j,k,0) to 0 at mglev=0 made the IAMR hotspot test
fail to converge

I believe that the smoothers are ok with sten > 0 but the BiCG is not.

On Mon, Mar 29, 2021 at 12:48 PM Weiqun Zhang ***@***.***>
wrote:

> Does IAMR have positive diagonals on mglev == 0 (i.e., original grids)?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AMReX-Codes/amrex/pull/1908#issuecomment-809664766>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRE6YTYVO2F2MI564ZFCHDTGDKRFANCNFSM42AGQFNQ>
> .
>


-- 
Ann Almgren
Senior Scientist; CCSE Group Lead

WeiqunZhang commented:
Interesting! I guess those are "non-physical" covered nodes.  Otherwise, it would be strange.  Maybe we only need to do this for the bottom solver,  `amrlev == 0 && mglev == NMGLevels(0)-1`?

asalmgren commented:
yes - they are nodes in the covered region but whose span reaches into the
un-covered region - and the sign only flips when the volfrac is small (in
the cases I've seen)

I believe we only need to do this if we're using BiCG -- both smoothing and
hypre seem to handle it ok - but this makes the logic more complicated

Given that the "fix" makes the smoother not converge though (in the IAMR
case), it appears we do need a more targeted fix




On Mon, Mar 29, 2021 at 12:57 PM Weiqun Zhang ***@***.***>
wrote:

> Interesting! I guess those are "non-physical" covered nodes. Otherwise, it
> would be strange. Maybe we only need to do this for the bottom solver, amrlev
> == 0 && mglev == NMGLevels(0)-1?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AMReX-Codes/amrex/pull/1908#issuecomment-809670015>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRE6YVAGZHMUFMJ4TMSHVLTGDLR7ANCNFSM42AGQFNQ>
> .
>


-- 
Ann Almgren
Senior Scientist; CCSE Group Lead

asalmgren commented:
It's also the case that I've only seen this in 3D, so we may want to only
add this fix for 3D until we see a 2D case fail ...

On Mon, Mar 29, 2021 at 1:04 PM asalmgren ***@***.***> wrote:

> yes - they are nodes in the covered region but whose span reaches into the
> un-covered region - and the sign only flips when the volfrac is small (in
> the cases I've seen)
>
> I believe we only need to do this if we're using BiCG -- both smoothing and
> hypre seem to handle it ok - but this makes the logic more complicated
>
> Given that the "fix" makes the smoother not converge though (in the IAMR
> case), it appears we do need a more targeted fix
>
>
>
>
> On Mon, Mar 29, 2021 at 12:57 PM Weiqun Zhang ***@***.***>
> wrote:
>
> > Interesting! I guess those are "non-physical" covered nodes. Otherwise,
> it
> > would be strange. Maybe we only need to do this for the bottom solver,
> amrlev
> > == 0 && mglev == NMGLevels(0)-1?
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/AMReX-Codes/amrex/pull/1908#issuecomment-809670015>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ACRE6YVAGZHMUFMJ4TMSHVLTGDLR7ANCNFSM42AGQFNQ
> >
> > .
> >
>
>
> --
> Ann Almgren
> Senior Scientist; CCSE Group Lead
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AMReX-Codes/amrex/pull/1908#issuecomment-809674233>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRE6YXNAKHLZQCZSTAU7SDTGDML3ANCNFSM42AGQFNQ>
> .
>


-- 
Ann Almgren
Senior Scientist; CCSE Group Lead

asalmgren committed:
Make the "fix" much less intrusive:
Only set the center coefficient to 0 if 1) positive 2) 3d, 2) amrlev == 0 and mglev == max mg level

asalmgren committed:
Don't put if test inside kernel
